docs/proposals/Periodic-Retrieval-of-Certificates-and-Keys-from-AKV.md

```diff
@@ -0,0 +1,121 @@
+
+### Certificate revocation
+
+A malicious user compromised specific versions of certificate. Alice promptly reported this incident to the Certificate Authority (CA). The CA added the compromised versions to the Certificate Revocation List (CRL) and responded to Online Certificate Status Protocol (OCSP) queries with the revoked status. As a result, images signed with the revoked versions fail signature verification. Alice rotated the certificate, obtaining a new version issued by the CA while maintaining the same certificate name and identity information. The build pipeline was triggered to build and sign all the images with the new version of the certificate. Furthermore, as Alice configures Ratify to periodically retrieve all available versions of the certificate from AKV, it ensures successful verification of images signed with the latest version without requiring manual configuration adjustments. By following these actions, Alice maintains security while seamlessly transitioning to the new certificate version.
+
+### Key rotation
+
+Bob, an application engineer at Wabbit Network LLC, focuses on securing containerized applications. He sets up the build pipeline to sign container images using Cosign and key pairs in AKV. To ensure that only trusted images are deployed in Kubernetes clusters, he deploys Gatekeeper and Ratify. Ratify is configured to validate Cosign signatures using keys from AKV. As a security best practice, Bob configured automated key rotation for the key in AKV, resulting a new version of the key created before key expiry. He ensures that the pipeline always uses the latest key version for signing container images. Additionally, Bob configures Ratify to periodically retrieve all available versions of the key from AKV, allowing successful verification of images signed by different key versions without any manual configuration adjustments. Bob knows that images are not allowed to be signed with Cosign using keys that have expired in AKV, but verification using expired key is allowed. He assumes that the old versions of the keys are retrieved by Ratify, because there are still some images around that were signed with old versions.
+
+### Key disabling
+
+A malicious user compromised the private key used for signing images. Bob promptly rotated the key, obtaining a new version, and disabled the compromised version. Bob configures Ratify to periodically retrieve all available versions of the key from AKV excluding any disabled versions, as a result, images signed with the disabled version fail signature verification. The build pipeline was triggered to build and sign all the images with the new version of key. As Ratify periodically retrieve all available versions of the key from AKV, it ensures successful verification of images signed with the latest key version without requiring manual configuration adjustments. By following these actions, Bob maintains security while seamlessly transitioning to the new version of key.
```

Comment:
Tying back to my previous question about "all previous versions", Ratify doesn't support this scenario unless the user defined each old version manually. Is this a new functional requirement?

Reply:
Would be good to capture what CR spec looks like for "1. fetch pinned version , 2. fetch all version , 3. fetch only latests"

Reply:
> Tying back to my previous question about "all previous versions", Ratify doesn't support this scenario unless the user defined each old version manually. Is this a new functional requirement?

I think this is exactly what the requirement, because if you always keep the latest one, you may fail to validate the images signed by old images, although after some time, all the images are signed by the latest one, but there will be a period that some images are signed with old version, some images are signed with new version. This is especially the case for the Cosign public key scenarios.

Reply:
> Would be good to capture what CR spec looks like for "1. fetch pinned version , 2. fetch all version , 3. fetch only latests"

As a best practice, key and cert are automated rotated, so all version matters, especially for Cosign public key scenarios.